src/main/java/net/neoforged/neoforge/client/model/generators/template/ExtendedModelTemplateBuilder.java

```diff
@@ -0,0 +1,178 @@
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder transforms(Consumer<TransformsBuilder> action) {
+        action.accept(transforms);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder ao(boolean ao) {
+        this.ambientOcclusion = ao;
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder guiLight(UnbakedModel.GuiLight light) {
+        this.guiLight = light;
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder element(Consumer<ElementBuilder> action) {
+        Preconditions.checkState(
+                customLoader == null || customLoader.allowInlineElements,
+                "Custom model loader %s does not support inline elements",
+                customLoader != null ? customLoader.loaderId : null);
+        ElementBuilder ret = new ElementBuilder();
+        action.accept(ret);
+        elements.add(ret);
+        return this;
+    }
+
+    /**
+     * Get an existing element builder
+     *
+     * @param index the index of the existing element builder
+     * @return the element builder
+     * @throws IndexOutOfBoundsException if {@code} index is out of bounds
+     */
+    public ExtendedModelTemplateBuilder element(int index, Consumer<ElementBuilder> action) {
+        Preconditions.checkState(
+                customLoader == null || customLoader.allowInlineElements,
+                "Custom model loader %s does not support inline elements",
+                customLoader != null ? customLoader.loaderId : null);
+        Preconditions.checkElementIndex(index, elements.size(), "Element index");
+        action.accept(elements.get(index));
+        return this;
+    }
+
+    /**
+     * {@return the number of elements in this model builder}
+     */
+    public int getElementCount() {
+        return elements.size();
+    }
+
+    /**
+     * Use a custom loader instead of the vanilla elements.
+     *
+     * @param customLoaderFactory function that returns the custom loader to set, given this
+     * @return the custom loader builder
+     */
+    public <L extends CustomLoaderBuilder> ExtendedModelTemplateBuilder customLoader(Supplier<L> customLoaderFactory, Consumer<L> action) {
+        Preconditions.checkNotNull(customLoaderFactory, "customLoaderFactory must not be null");
+        L customLoader = customLoaderFactory.get();
+        Preconditions.checkState(
+                customLoader.allowInlineElements || elements.isEmpty(),
+                "Custom model loader %s does not support inline elements",
+                customLoader.loaderId);
+        this.customLoader = customLoader;
+        action.accept(customLoader);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder rootTransforms(Consumer<RootTransformsBuilder> action) {
+        action.accept(rootTransforms);
+        return this;
+    }
+
+    public ExtendedModelTemplate build() {
+        return new ExtendedModelTemplate(this);
+    }
```

Comment:
These methods need short doc comments, particularly the root transforms one to differentiate it from the other transforms method.